libbzip2-rs-sys/src/bzlib.rs

```diff
@@ -483,28 +513,18 @@ unsafe fn configure_allocator(strm: *mut bz_stream) -> Option<Allocator> {
         }
         (None, None) => {
             let allocator = Allocator::DEFAULT?;
-            let (bzalloc, bzfree) = allocator.function_pointers();
+            let (bzalloc, bzfree) = Allocator::default_function_pointers()?;
 
             (*strm).bzalloc = Some(bzalloc);
             (*strm).bzfree = Some(bzfree);
 
             Some(allocator)
         }
-        _ => {
-            // this is almost certainly a bug, but replicates the original C behavior.
-            //
-            // Note that this logic does not really work with the default rust allocator, because
-            // it will panic at runtime when called directly. Usually the idea here is that
-            // allocation is special, and free is just the default `libc::free` that we configure
-            // by default with the default C allocator.
-            let allocator = Allocator::DEFAULT?;
-            let (default_bzalloc, default_bzfree) = allocator.function_pointers();
-
-            let bzalloc = (*strm).bzalloc.get_or_insert(default_bzalloc);
-            let bzfree = (*strm).bzfree.get_or_insert(default_bzfree);
-
-            Some(Allocator::custom(*bzalloc, *bzfree, (*strm).opaque))
-        }
+        // Using a different allocator for alloc and free is UB. The user of libbzip2-rs can't get a
+        // reference to the default alloc or free function, so hitting this path means that using
+        // the default alloc or free function would cause two allocators to be mixed. As such return
+        // an error to prevent UB.
```

Comment:
it is documented behavior that the default (C) allocator uses malloc/free, so it is totally possible to provide just a valid alloc function (ultimately calling malloc, calloc, maybe doing some extra bookkeeping) and rely on the dealloc function being `libc::free`. 

Do we really gain anything here? we already assume that the allocation function pointers are correct, responsibility is explicitly on the caller.

Reply:
The default allocator for the rust version is not malloc/free when the rust-allocator feature is enabled.

Reply:
that's true but again, the responsibility for providing correct function pointers is explicitly on the caller here. Is it really necessary to break compatibility with the C version here? do we gain anything real?

Reply:
We already break compatibility with the C version when rust-allocator is enabled. This ensures that things don't silently go wrong when the caller assumed that it is fine to only set bzalloc or bzfree as would have been the case in the C version.

Reply:
yes but this change also breaks compatibility when just the C allocator is enabled right?

Reply:
Technically yes, in practice I did say the code was wrong in the first place.

Reply:
I agree, but that's not up to us though. drop-in compatible has consequences, and [the manual](https://www.sourceware.org/bzip2/manual/manual.html) is just too vague here, and I could totally see some cursed old code doing this. So we can deviate if the default allocator is rust, but not when it is C.

Reply:
I haven't found a single misuse across the entirety of github.

Reply:
that also means that either option has the same effect, right? (also are many people even using custom allocators? I think the cases that do are almost by definition kind of cursed)

What I worry about is that in theory this bzip2 could get packaged by fedora/ubuntu,  it makes its way to some strange place, and there an application breaks because we made a change. A change that we did not really need to make, and had no real benefit. Why run that risk?